src/index.ts

```diff
@@ -0,0 +1,208 @@
+/**
+ * @fileoverview This is an ES Module which exports `vizObserver`, see its function below.
+ *
+ * This uses a combination of IntersectionObserver and ResizeObserver to inform you when DOM
+ * nodes move around on a page (and also when they're resized). This might be useful if
+ * you're trying to track a complex HTML structure, perhaps one you don't control.
+ *
+ * It also works with target nodes inside Shadow DOM.
+ */
+
+/**
+ * This is an element's position relative to the page's documentElement.
+ */
+export interface Rect {
+  appear: boolean;
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+}
+
+export interface Options {
+  signal?: AbortSignal;
+}
+
+export interface VisualObserverEntry {
+  target: Element;
+  contentRect: DOMRectReadOnly;
+  isAppearing: boolean;
+}
+
+export interface VisualObserverCallback {
+  (entries: VisualObserverEntry[], observer: VisualObserver): void;
+}
+
+interface VisualObserverElement {
+  intersectionObserver: IntersectionObserver;
+}
+
+/**
+ * Create an observer that notifies when an element resizes, moves, or is added/removed from the DOM.
+ */
+export class VisualObserver {
+  #callback: VisualObserverCallback;
+
+  constructor(callback: VisualObserverCallback) {
+    this.#callback = callback;
+  }
+
+  #elementCount = 0;
+  // Unsure whether to use a Map or WeakMap. Not using a weak map means we are prone to memory leaks, but it also means that we cant iterate and/or clear the map.
+  #elements = new WeakMap<Element, VisualObserverElement>();
```

Comment:
Suck on whether to use a WeakMap here. Not sure how to implement `disconnect` with it...

Reply:
Think we just need to go with `Map` for the time being...